numalogic/registry/redis_registry.py

```diff
@@ -0,0 +1,247 @@
+        )
+        return new_version_key
+
+    def load(
+        self,
+        skeys: Sequence[str],
+        dkeys: Sequence[str],
+        production: bool = True,
+        version: str = None,
+    ) -> Optional[ArtifactData]:
+        """
+        Saves the artifact into mlflow registry and updates version.
+        Args:
+            skeys: static key fields as list/tuple of strings
+            dkeys: dynamic key fields as list/tuple of strings
+            production: load the model in production stage
+            version: version to load
+
+        Returns:
+            mlflow ModelVersion instance
+        """
+        try:
+            if (production and version) or (not production and not version):
+                raise ValueError("Either One of 'latest' or 'version' needed in load method call")
+            model_key = self.__get_model_key(production, version, skeys, dkeys)
+            serialized_model, model_version, model_timestamp = self.client.hmget(
+                name=model_key, keys=["model", "version", "timestamp"]
+            )
+            deserialized_model = loads(serialized_model)
+            metadata = self.__load_metadata(model_key)
+            return ArtifactData(
+                artifact=deserialized_model,
+                metadata=metadata,
+                extras={
+                    "model_timestamp": int(model_timestamp.decode()),
+                    "model_version": model_version.decode(),
+                },
+            )
+        except ModelKeyNotFound as model_key_error:
+            _LOGGER.exception("Missing Key: %s", model_key_error)
+            return None
+        except RedisError as ex:
+            _LOGGER.exception("Unexpected error: %s", ex)
+            return None
+
+    def save(
+        self,
+        skeys: Sequence[str],
+        dkeys: Sequence[str],
+        artifact: Artifact,
+        **metadata: str,
+    ) -> Optional[str]:
+        """
+        Saves the artifact into redis registry and updates version.
+        Args:
+            skeys: static key fields as list/tuple of strings
+            dkeys: dynamic key fields as list/tuple of strings
+            artifact: primary artifact to be saved
+            metadata: additional metadata surrounding the artifact that needs to be saved
+
+        Returns:
+            model version
+        """
+        try:
+            production_key = self.construct_key(skeys, dkeys, ["PRODUCTION"])
+            version = 0
+            if self.client.exists(production_key):
+                _LOGGER.info("Production key exists for the model")
+                version_key = self.client.get(name=production_key)
+                version = int(self.get_version(version_key.decode())) + 1
+            new_version_key = self.__save_artifact(artifact, skeys, dkeys, str(version))
+            if metadata:
+                self.__save_metadata(metadata, new_version_key)
+            self.client.expire(name=new_version_key, time=self.ttl)
```

Comment:
Do we apply 3 separate operations to save an artifact?
1. save key value
2. save production
3. set expire

When the pod fails over or network blips in-between 1 and 2, or 2 and 3, we will leak entries in Redis forever.
All three operations should be done in one write to Redis.

Reply:
Further more, either 5000 or 50000 seconds TTL seems too short for model storage.

Reply:
Yes. I have changed it to 5days, ie: 604800 fir now.

Also, have added redis pipeline to solve the use case here. It will execute all the three commands now sequentially.